app/models/manageiq/providers/amazon/inventory/persister/target_collection.rb

```diff
@@ -1,145 +1,148 @@
-    )
-
-    add_inventory_collection_with_references(
-      network,
-      :floating_ips,
-      references(:floating_ips) + references(:load_balancers),
-      :parent => manager.network_manager
-    )
-
-    # Child models with references in the Parent InventoryCollections for Network
-    add_inventory_collections(
-      network,
-      %i(firewall_rules cloud_subnet_network_ports load_balancer_pools load_balancer_pool_members
-         load_balancer_pool_member_pools load_balancer_listeners load_balancer_listener_pools
-         load_balancer_health_checks load_balancer_health_check_members),
-      :parent   => manager.network_manager
-    )
-
-    ######### Storage ##############
-    # Top level models with direct references for Network
-    add_inventory_collections_with_references(
-      storage,
-      %i(cloud_volumes cloud_volume_snapshots),
-      :parent => manager.ebs_storage_manager
-    )
+    initialize_cloud_inventory_collections
 
-    if manager.s3_storage_manager
-      add_inventory_collections_with_references(
-        storage,
-        %i(cloud_object_store_containers cloud_object_store_objects),
-        :parent => manager.s3_storage_manager
-      )
-    end
+    initialize_network_inventory_collections
 
-    # Model we take just from a DB, there is no flavors API
-    add_inventory_collections(
-      cloud,
-      %i(flavors),
-      :strategy => :local_db_find_references
-    )
-
-    ######## Custom processing of Ancestry ##########
-    add_inventory_collection(
-      cloud.vm_and_miq_template_ancestry(
-        :dependency_attributes => {
-          :vms           => [collections[:vms]],
-          :miq_templates => [collections[:miq_templates]]
-        }
-      )
-    )
-
-    add_inventory_collection(
-      cloud.orchestration_stack_ancestry(
-        :dependency_attributes => {
-          :orchestration_stacks           => [collections[:orchestration_stacks]],
-          :orchestration_stacks_resources => [collections[:orchestration_stacks_resources]]
-        }
-      )
-    )
+    initialize_storage_inventory_collections
   end
 
   private
 
-  def add_inventory_collections_with_references(inventory_collections_data, names, options = {})
-    names.each do |name|
-      add_inventory_collection_with_references(inventory_collections_data, name, references(name), options)
-    end
+  def initialize_cloud_inventory_collections
+    init_cloud_ics_top_level_models
+
+    init_cloud_ics_child_models
+
+    # Custom processing of Ancestry
+    add_vm_and_miq_template_ancestry
+
+    add_orchestration_stack_ancestry
   end
 
-  def add_inventory_collection_with_references(inventory_collections_data, name, manager_refs, options = {})
-    options = shared_options.merge(inventory_collections_data.send(
-      name,
-      :manager_uuids => manager_refs,
-    ).merge(options))
+  # Top level models with direct references for Cloud
+  def init_cloud_ics_top_level_models
```

Comment:
Hm, instead of top/child level separation, I would rather like to see these shared with ManageIQ::Providers::Amazon::Inventory::Persister::CloudManager and ManageIQ::Providers::Amazon::Inventory::Persister::NetworkManager (and storage)

The only difference is that the TargetCollection is passing references(XY) to few collections. And I think we can even pass the references everytime. Because they are not used if we pass ```:targeted => false```, which is extracetd to method above and should be passed to every IC

Reply:
right, so you can still pass the references only optionally. The point is sharing the definitions, which will help the readibility and will make it more composable.